src/bse/K_stored_in_a_nest_matrix.F

```diff
@@ -0,0 +1,337 @@
+! be useful, but WITHOUT ANY WARRANTY; without even the 
+! implied warranty of MERCHANTABILITY or FITNESS FOR A 
+! PARTICULAR PURPOSE.  See the GNU General Public License 
+! for more details.
+!
+! You should have received a copy of the GNU General Public 
+! License along with this program; if not, write to the Free 
+! Software Foundation, Inc., 59 Temple Place - Suite 330,Boston, 
+! MA 02111-1307, USA or visit http://www.gnu.org/copyleft/gpl.txt.
+!
+
+module shellmat_module
+#include <petsc/finclude/petscmat.h>
+  use petscmat
+  Mat::R,C
+end module shellmat_module
+
+subroutine ignore_petsc_error_handler(comm,line,fun,file,n,p,mess,ctx,ierr)
+!
+! Handler to ignore error if MatMultHermitianTranspose cannot be set.
+! This operation is only available for shell matrices in PETSc from v.3.21
+! Setting MatMultTranspose is enough, but it is preferable to set both.
+!
+  use petscsysdef
+  integer line,n,p
+  PetscInt ctx 
+  PetscErrorCode ierr
+  MPI_Comm comm
+  character*(*) fun,file,mess
+end subroutine ignore_petsc_error_handler
+
+subroutine K_stored_in_a_nest_matrix(i_BS_mat,slepc_mat)
+!
+! BS_K_coupling and l_BS_ares_from_res are assumed TRUE
+!
+! K is stored in a PETSc nest matrix, taking advantage of the four block
+! structure of the matrix in this case:
+!
+! | R     C  |
+! |-C^*  -R^T|
+!
+#include <petsc/finclude/petscsys.h>
+#include <petsc/finclude/petscvec.h>
+#include <petsc/finclude/petscmat.h>
+#include <petsc/finclude/petscvec.h>
+#include <slepc/finclude/slepcsys.h>
+#include <slepc/finclude/slepceps.h>
+!
+  use pars,           ONLY:cI,cONE
+  use BS,             ONLY:BS_K_dim,BS_H_dim,BS_blk,n_BS_blks
+  use BS_solvers,     ONLY:BSS_eh_E,BSS_eh_W,BSS_perturbative_width
+  use cuda_m,         ONLY:have_cuda
+  !
+  use petscmat
+  use slepceps
+  use slepcepsdef
+  use petscmatdef
+  use shellmat_module
+  !
+  implicit none
+  !
+  integer, intent(in)  :: i_BS_mat
+  Mat,     intent(out) :: slepc_mat
+  Mat                  :: RT,CHT
+  Mat                  :: matArray(4)
+  !
+  integer     :: i_c,i_r,i_Tk,i_Tp,i_B,H_shift(2)
+  PetscScalar         :: Mij,Mij_star
+  PetscScalar         :: mone
+  PetscInt            :: H_pos(2),SL_K_dim(2),SL_H_dim
+  PetscInt            :: itwo
+  PetscErrorCode      :: ierr
+  !
+  external RT_mat_mult,CHT_mat_mult,& 
+           RT_mat_mult_transpose,CHT_mat_mult_transpose,&
+           RT_mat_mult_hermitian_transpose,CHT_mat_mult_hermitian_transpose,&
+           RT_get_diagonal,CHT_get_diagonal,&
+           ignore_petsc_error_handler
+  !
+  SL_H_dim=BS_H_dim
+  SL_K_dim=BS_K_dim
+  !
+  ! Allocate the explicit submatrices of the nest matrix
+  !
+  call MatCreate(PETSC_COMM_WORLD,R,ierr)
+  call MatSetSizes(R,PETSC_DECIDE,PETSC_DECIDE,SL_H_dim/2,SL_H_dim/2,ierr)
+  call MatSetType(R,MATMPIDENSE,ierr)
+  if (have_cuda) then
+#ifdef PETSC_HAVE_CUDA
+  call MatSetType(R,MATDENSECUDA,ierr)
+  call MatSetVecType(R,VECCUDA,ierr)
```

Comment:
This should not be necessary, since it is already the default `VecType` in `MATDENSECUDA`. It is only necessary for shell matrices and `MATNEST`.
```suggestion
```